xformers/components/attention/blocksparse.py

```diff
@@ -169,13 +191,38 @@ def forward(
             # Blocksparse only works on fp16
             q_dtype = q.dtype
             q, k, v = q.half(), k.half(), v.half()
+            if att_mask is not None:
+                att_mask = att_mask.half()
 
             # Self-attend: (B, nh, S, hs) x (B, nh, hs, S) -> (B, nh, S, S)
             # When the computations are block sparse, the matrix types change along the way:
             # - (sparse) attention matrix = (dense) Kt * (dense) Q
             q = q / math.sqrt(q.size(-1))
             sparse_att_mat = self.sparse_dot_sdd(q, k)
 
+            # apply masks
+            if att_mask is not None:
+                # mask shape is either (B,1,1,S) or (B,nh,1,S)
```

Comment:
in other parts of xformers (and more generally I think given their name) attention masks relate to the attention matrix, so they don't only mark key or query but are (B,S,S) or (S,S). would that work for you ?